### PR TITLE
Ifpack2: Add Schwarz-like Database Preconditioner

### PR DESCRIPTION
**Pull request by GrahamBenHarper**
@trilinos/ifpack2 

## Motivation
Patch-based preconditioning methods generally reduce the iteration count considerably for high-order and problems with local dense couplings. However, in these types of problems there tends to be a large amount of inherent structure on the mesh level. Define the restriction of the matrix $A$ to each patch by $A_k, k=1,\dots,n_p$ where each patch is a $p_s \times p_s$ matrix. This preconditioner computes a greedy compressed representation on each patch $A_k$ by checking $\| A_k - B \|_{\ell_1} < \varepsilon$ for every $B$ in a database. $A_k$ is assigned an index to the first $B$ satisfying such, and if no such $B$ is found, $A_k$ is appended to the list of $B$ matrices. For many high-order problems, this tends to compress things by 95% to 99% without loss of accuracy. This compression increases data locality and accelerates the solver at a small setup cost.

This preconditioner assumes:
1. The locally owned portion of $A$ contains all locally owned patches. This is accomplished by using the `AdditiveSchwarz` class to construct a distance-1 overlapped matrix $A$ beforehand.
2. For each patch, there is at least one row in the matrix $A$ with $p_s$ degrees of freedom.
3. The union of all degrees of freedom corresponding to such patches covers all degrees of freedom.

### CODE REVIEW

File: packages/ifpack2/test/unit_tests/CMakeLists.txt
Comment (cgcgcg) [inline packages/ifpack2/test/unit_tests/CMakeLists.txt:90]: Since it took some work to get this working for several ranks, maybe test it also on 4 ranks?

File: packages/ifpack2/test/unit_tests/Ifpack2_UnitTestDatabaseSchwarz.cpp
Comment (cgcgcg) [inline packages/ifpack2/test/unit_tests/Ifpack2_UnitTestDatabaseSchwarz.cpp:162]: Actually test the result?

File: packages/muelu/test/scaling/CMakeLists.txt
Comment (cgcgcg) [inline packages/muelu/test/scaling/CMakeLists.txt:336]: Same deal.

File: packages/muelu/test/scaling/DatabaseSchwarzSmoo.xml
Comment (cgcgcg) [inline packages/muelu/test/scaling/DatabaseSchwarzSmoo.xml:36]: "9 DoFs per patch for degree 2 in 2D on quads"

File: packages/muelu/test/scaling/DatabaseSchwarzSmoo.xml
Comment (cgcgcg) [inline packages/muelu/test/scaling/DatabaseSchwarzSmoo.xml:21]: Why is this used as a coarse solver instead of a smoother?

File: packages/ifpack2/src/Ifpack2_DatabaseSchwarz_decl.hpp
Comment (cgcgcg) [inline packages/ifpack2/src/Ifpack2_DatabaseSchwarz_decl.hpp:102]: Would it make sense to derive from Ifpack2::AdditiveSchwarz (or whatever this is based on) instead of duplicating?

File: packages/ifpack2/src/Ifpack2_DatabaseSchwarz_def.hpp
Comment (cgcgcg) [inline packages/ifpack2/src/Ifpack2_DatabaseSchwarz_def.hpp:240]: WIP?

**Review by cgcgcg — COMMENTED** [review]
What was the motivation to write all of this using host-only methods? Wouldn't it have made more sense to directly write it using Kokkos?

File: packages/ifpack2/test/unit_tests/Ifpack2_UnitTestDatabaseSchwarz.cpp
Comment (GrahamBenHarper) [reply packages/ifpack2/test/unit_tests/Ifpack2_UnitTestDatabaseSchwarz.cpp:162]: I don't have any good examples of analytic solutions that I can port into this testing framework without using something like a gold file

Comment (GrahamBenHarper) [reply packages/ifpack2/test/unit_tests/Ifpack2_UnitTestDatabaseSchwarz.cpp:162]: I was thinking one good test is the `"skip database" = true` route should agree with the `"skip database" = false` route

File: packages/muelu/test/scaling/CMakeLists.txt
Comment (GrahamBenHarper) [reply packages/muelu/test/scaling/CMakeLists.txt:336]: I can definitely do that here since I have it wrapped in the `AdditiveSchwarz` solver

File: packages/ifpack2/src/Ifpack2_DatabaseSchwarz_def.hpp
Comment (GrahamBenHarper) [reply packages/ifpack2/src/Ifpack2_DatabaseSchwarz_def.hpp:240]: Good catch! That should be false

File: packages/ifpack2/src/Ifpack2_DatabaseSchwarz_decl.hpp
Comment (GrahamBenHarper) [reply packages/ifpack2/src/Ifpack2_DatabaseSchwarz_decl.hpp:102]: The options I looked at were 
 1. Add this as another option in `AdditiveSchwarz`.
 2. Derive from `AdditiveSchwarz`.
 3. Derive from `Preconditioner`, and use it as an inner solve with `AdditiveSchwarz`.

For 1., it seemed like this main idea of constructing a database was large enough and required enough code and data structures that this wouldn't make sense. For 2., I could do something like that and then force this to be the inner solve every time, but there would be weird conflicts like having to rewrite the parameterlist settings with the new name and then trying to explicitly call the parent class methods first anyway. I felt like 3. was the easiest route since we already had options to handle inner preconditioners with `AdditiveSchwarz`, so I wouldn't have to manually juggle any parent-child interactions/reimplementations/overrides/etc.

**Comment by GrahamBenHarper** [comment]
@cgcgcg overall, the main reason was simply time. The project that funded this was much too short of a time period with too many tasks to accomplish, so a host-only implementation was the only thing we had time for. Originally there were people who suggested there would be follow-on opportunities, but none of those follow-on opportunities materialized. My main motivation now is to merge this so that it can get some visibility and hopefully find people with enough interest to support additional options or improvements for it. I'd definitely love to find more time to improve this and move it over to using Kokkos and KokkosKernels under the hood.

At the moment, there are quite a lot of Stokhos build errors due to it trying to use Sacado AD types with this, which isn't something I even considered before, so it looks like I have a bit more work before I can get this merged.

File: packages/muelu/test/scaling/DatabaseSchwarzSmoo.xml
Comment (GrahamBenHarper) [reply packages/muelu/test/scaling/DatabaseSchwarzSmoo.xml:21]: I didn't actually want to run multigrid because a lot of the problems I tested were high-order, and my understanding was if I wanted a smoother-only test then I needed `"coarse: type"` with a smoother in it, similar to `oLabSchwarzEquivBlkSmoo.xml` and `generalBlkSmoothing.xml`. With that said, the repartitioning options below can definitely be removed.

Comment (cgcgcg) [reply packages/muelu/test/scaling/DatabaseSchwarzSmoo.xml:21]: The smoother-only behavior can be tested in Ifpack2. Behavior in a multigrid solver can be tested in MueLu.

File: packages/ifpack2/test/unit_tests/Ifpack2_UnitTestDatabaseSchwarz.cpp
Comment (cgcgcg) [reply packages/ifpack2/test/unit_tests/Ifpack2_UnitTestDatabaseSchwarz.cpp:162]: I'd just cook up some simple problem with a known solution and a couple of patches.

File: packages/muelu/test/scaling/DatabaseSchwarzSmoo.xml
Comment (GrahamBenHarper) [reply packages/muelu/test/scaling/DatabaseSchwarzSmoo.xml:21]: Do we have drivers in Ifpack2 for loading matrix-market data?

Comment (cgcgcg) [reply packages/muelu/test/scaling/DatabaseSchwarzSmoo.xml:21]: Yes

File: packages/ifpack2/src/Ifpack2_DatabaseSchwarz_def.hpp
Comment (cgcgcg) [inline packages/ifpack2/src/Ifpack2_DatabaseSchwarz_def.hpp:343]: Here?

**Comment by GrahamBenHarper** [comment]
Okay I fixed the type errors with complex builds and Sacado AD types in Stokhos, but I still need to improve the robustness of the tests to make sure it's behaving properly.

File: packages/ifpack2/test/unit_tests/Ifpack2_UnitTestDatabaseSchwarz.cpp
Comment (GrahamBenHarper) [reply packages/ifpack2/test/unit_tests/Ifpack2_UnitTestDatabaseSchwarz.cpp]: Okay I figured it out. I'm not sure what the default matrix is, but I got an analytic solution, so I'm happy enough

**Comment by GrahamBenHarper** [comment]
Okay I believe I have everything polished at this point. I've verified input decks work with the Belos test, I fixed the unit test, I fixed an issue where it wasn't using the multivector properly, I added some more verbose error checking so that it will error out and tell the user if it isn't going to work, and I removed the tests from MueLu for now at @cgcgcg's request since I don't have a good excuse/example to use it with MueLu at the moment.

**Comment by GrahamBenHarper** [comment]
Blocked by #11621

**Review by cgcgcg — APPROVED** [review]

**Comment by GrahamBenHarper** [comment]
A single test failed. `PR-11517-test-rhel7_sems-intel-2021.3-sems-openmpi-4.0.5_release-debug_shared_no-kokkos-arch_no-asan_no-complex_fpic_mpi_no-pt_no-rdc_no-uvm_deprecated-on_no-package-enables-344` test name `Stratimikos_Galeri_xpetra_double_Jacobi_Tpetra_MPI_4` reports
```
[ascic142][[26845,1],1][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],0][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],0][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],1][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(17) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],2][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
[ascic142][[26845,1],3][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(16) failed: Connection reset by peer (104)
```

**Comment by GrahamBenHarper** [comment]
Thank you @cgcgcg!

Also @sebrowne do you know if this was some spurious network hiccup or maintenance that caused the error with testing in https://github.com/trilinos/Trilinos/pull/11517#issuecomment-1528844100 or if it was potentially something else?

**Comment by sebrowne** [comment]
@GrahamBenHarper at a glance yes, that's classic "system problem".  I don't know if the machine was undergoing maintenance (if it was expected), but yes I think it's what you said.  I checked it manually and it doesn't appear to have any extreme issues at a glance (though it's always tough to say that for certain).

**Comment by GrahamBenHarper** [comment]
Okay that's good to know. Thank you @sebrowne !

**Pull request merged**
